### PR TITLE
Alias for a simplified BCP parameterisation 

### PR DESCRIPTION
**Pull request by rrsadykov**
Introduces `Coluna.Algorithm.BranchCutAndPriceAlgorithm` alias for a simplified parameterisation of the branch-cut-and-price algorithm.

Addresses issue #478.

### CODE REVIEW

**Comment by rrsadykov** [comment]
Codecov Report is wrong

**Comment by guimarqu** [comment]
I think that before merging this new interface, we must decide on how we define the branching priorities.

My suggestion is to define branching priorities only through BlockDecomposition. Currently, we only have `BlockDecomposition.branchingpriority!` to define the branching priority on a JuMP variable. I think it can replace `VarBranchingRule`. However, if the latter disappears, `AbstractBranchingRule` and `PrioritisedBranchingRule` also disappear.

The aim of `AbstractBranchingRule` is not clear to me.

**Comment by rrsadykov** [comment]
`AbstractBranchingRule` is the algorithm type. Algorithms of this type find branching candidates. `VarBranchingRule` is a particular algorithm to perform branching. There can be other branching rules: Ryan&Foster, branching on constraints (defined by ``@expression` in JuMP), or the user can define his own branching rules. May be we should rename `BranchingRule` to `BranchingAlgorithm` to be more clear. 

The main problem here is that we mix variable branching priorities given using `BlockDecomposition` and branching algorithms defined in Coluna (these two parameterisations should be "compatible"). We cannot have just variable branching priorities, as it is important to be able to do other kinds of branching (non-robust branching is sometimes necessary for completeness).   

One solution may be to define variable branching using names, like 

```
branchingFirstOnXthenOnY = ClA.StrongBranching(
            phases = [] ,
            rules = [ClA.PrioritisedBranchingRule(ClA.VarBranchingRule(varname = "X"),  2.0, 2.0),
                         ClA.PrioritisedBranchingRule(ClA.VarBranchingRule(varname = "Y"), 1.0, 1.0)]
        )
```

For this, we need to be able to get the JuMP name of a variable (or expression) inside the branching algorithm in Coluna.

Then, branching priorities given using `BlockDecomposition` will serve just to sort individual variables with the same name. So, the "upper level" branching priorities will be given in the Coluna parameterisation, and "lower level" branching priorities will be given using `BlockDecomposition`. Actually, this is completely compatible with the current implementation. 

One may also parameterise `VarBranchingRule` with the selection strategy (most violated, smallest index, etc). This is how it is done in constraint programming solvers.

**Comment by guimarqu** [comment]
Ok, thanks for the explanation. I had doubts about the aim of `AbstractBranchingRule` because there is currently only one rule implemented & because the selection criterion is in `StrongBranching`.

About your suggestion, I don't think we can rely on MathOptInterface to always provide us the name of the variables and constraints. Currently, if you build your model using the caching optimizer (which is the default mode), Coluna does not receive the name of variables and constraints from MathOptInterface. 
Moreover, users may want to use [bridges](https://jump.dev/MathOptInterface.jl/v0.10/submodules/Bridges/overview/) to transform variables and constraints unsupported by Coluna into supported ones.
I guess that some bridges may create new variables but how do you propagate the branching rules to these new variables with your suggestion ? I think that these new variables have no name.

**Comment by rrsadykov** [comment]
Receiving names of variables (and constraints) is critical. In many cases, the user needs to know on which variable the branching is performed. 

In the worst case the name should be passed through `BlockDecomposition`. One may add variable name parameter in `BlockDecomposition.branchingpriority!()` or add another function `BlockDecomposition.varname!()`.

**Comment by rrsadykov** [comment]
> Moreover, users may want to use [bridges](https://jump.dev/MathOptInterface.jl/v0.10/submodules/Bridges/overview/) to transform variables and constraints unsupported by Coluna into supported ones.
> I guess that some bridges may create new variables but how do you propagate the branching rules to these new variables with your suggestion ? I think that these new variables have no name.

This is too advanced for me. Created variables may just receive some default (f.e. empty) name. The user then knows that branching is done on a variables not defined by him.

**Comment by guimarqu** [comment]
I agree that showing names of variables in branching is very useful when you implement a new application or in the R&D phase. However, I don't think we need them in production, so it's not so critical nor a requirement.
 
I guess created variables receive an empty name. However, I think we should be able to propagate the branching rule defined on the original variables because the latter may disappear. This propagation is done by MathOptInterface with a bridge that we will define for the BlockDecomposition.BranchingPriority attribute.
If we define branching rules in the algorithm, we can't do the propagation anymore.

I think it's very important to not try to bypass what has been done in MathOptInterface because we'll gain nothing except new bugs. The name should be provided by JuMP and MathOptInterface, not by BlockDecomposition.

**Comment by rrsadykov** [comment]
> However, I don't think we need them in production, so it's not so critical nor a requirement.

In research, there is not production phase. 

> The names should be provided by JuMP and MathOptInterface, not by BlockDecomposition

I completely agree, but you said that we cannot count on that. 

I general, I do not understand what do you suggest to do with branching.

**Comment by guimarqu** [comment]
I think we can define both branching priorities and branching rules through BlockDecomposition

```julia
branchingpriority!(x; rule = Coluna.VarBranchingRule(), root_priority = 1, non_root_priority = 2)
```

Before starting the optimization, Coluna can group variables that have the same branching rules and names are not a requirement anymore.

File: src/Algorithm/branchcutprice.jl [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+        opt_atol::Float64 = Coluna.DEF_OPTIMALITY_ATOL,
+        opt_rtol::Float64 = Coluna.DEF_OPTIMALITY_RTOL,
+        restmastipheur_timelimit::Int = 600,
+        restmastipheur_frequency::Int = 1,
+        restmastipheur_maxdepth::Int = 1000,
+        max_nb_cut_rounds::Int = 3,
+        colgen_stabilization::Float64 = 0.0, 
+        colgen_cleanup_threshold::Int = 10000,
+        colgen_stages_pricing_solvers::Vector{Int} = [1],
+        stbranch_phases_num_candidates::Vector{Int} = Vector{Int}(),
```
Comment (guimarqu) [inline src/Algorithm/branchcutprice.jl:14]:
```suggestion
        stbranch_phases_num_candidates::Vector{Int} = Int[],
```

File: src/Algorithm/branchcutprice.jl [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        stbranch_intrmphase_stages::Vector{NamedTuple{(:userstage, :solverid, :maxiters), Tuple{Int64, Int64, Int64}}}
+    )
+
+Coluna.Algorithm.BranchCutAndPriceAlgorithm is an alias for a simplified parameterisation
```
Comment (guimarqu) [inline src/Algorithm/branchcutprice.jl:18]:
```suggestion
Alias for a simplified parameterisation 
```

File: src/Algorithm/branchcutprice.jl [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        opt_atol::Float64 = Coluna.DEF_OPTIMALITY_ATOL,
+        opt_rtol::Float64 = Coluna.DEF_OPTIMALITY_RTOL,
+        restmastipheur_timelimit::Int = 600,
+        restmastipheur_frequency::Int = 1,
+        restmastipheur_maxdepth::Int = 1000,
+        max_nb_cut_rounds::Int = 3,
+        colgen_stabilization::Float64 = 0.0, 
+        colgen_cleanup_threshold::Int = 10000,
+        colgen_stages_pricing_solvers::Vector{Int64} = [1],
+        stbranch_phases_num_candidates::Vector{Int64} = Vector{Int64}(),
```
Comment (guimarqu) [inline src/Algorithm/branchcutprice.jl:67]:
```suggestion
        stbranch_phases_num_candidates::Vector{Int64} = Int[],
```

File: src/Algorithm/branchcutprice.jl [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+        stbranch_intrmphase_stages::Vector{NamedTuple{(:userstage, :solverid, :maxiters), Tuple{Int64, Int64, Int64}}} = [(userstage=1, solverid=1, maxiters=100)]
+)
+    heuristics::Vector{ParameterisedHeuristic} = []
```
Comment (guimarqu) [inline src/Algorithm/branchcutprice.jl:70]:
```suggestion
    heuristics = ParameterisedHeuristic[]
```

File: src/Algorithm/branchcutprice.jl [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+                get_dual_bound = false,
+                time_limit = restmastipheur_timelimit
+            )),
+            1.0, 1.0, restmastipheur_frequency, 
+            restmastipheur_maxdepth, "Restricted Master IP"
+        )
+        push!(heuristics, heuristic)
+    end
+
+    colgen_stages::Vector{ColumnGeneration} = []
```
Comment (guimarqu) [inline src/Algorithm/branchcutprice.jl:83]:
```suggestion
    colgen_stages = ColumnGeneration[]
```

File: src/Algorithm/branchcutprice.jl [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+    conquer = ColCutGenConquer(
+        stages = colgen_stages,
+        max_nb_cut_rounds = max_nb_cut_rounds,
+        primal_heuristics = heuristics,
+        opt_atol = opt_atol,
+        opt_rtol = opt_rtol
+    )
+
+    branching = NoBranching()
+    branching_rules::Vector{PrioritisedBranchingRule} = [PrioritisedBranchingRule(VarBranchingRule(), 1.0, 1.0)]
```
Comment (guimarqu) [inline src/Algorithm/branchcutprice.jl:112]:
```suggestion
    branching_rules = PrioritisedBranchingRule[PrioritisedBranchingRule(VarBranchingRule(), 1.0, 1.0)]
```

File: src/Algorithm/branchcutprice.jl [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+    if !isempty(stbranch_phases_num_candidates)
+        branching_phases::Vector{BranchingPhase} = []
```
Comment (guimarqu) [inline src/Algorithm/branchcutprice.jl:115]:
```suggestion
        branching_phases = BranchingPhase[]
```

File: src/Algorithm/branchcutprice.jl [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+        if length(stbranch_phases_num_candidates) >= 2
+            push!(branching_phases, 
+                BranchingPhase(first(stbranch_phases_num_candidates), RestrMasterLPConquer())
+            )    
+            if length(stbranch_phases_num_candidates) >= 3
+                intrmphase_stages::Vector{ColumnGeneration} = []
```
Comment (guimarqu) [inline src/Algorithm/branchcutprice.jl:121]:
```suggestion
                intrmphase_stages = ColumnGeneration[]
```

File: src/Algorithm/branching/varbranching.jl [hunk 9]
```diff
@@ -85,18 +85,30 @@ function run!(
     master = getmaster(reform)
     groups = Vector{BranchingGroup}()
     local_id = input.local_id
+    selected_vars = Vector{Pair{VarId, Float64}}()
```
Comment (guimarqu) [inline src/Algorithm/branching/varbranching.jl:88]:
```suggestion
    selected_vars = Pair{VarId, Float64}[]
```

File: src/Algorithm/branching/varbranching.jl [hunk 10]
```diff
@@ -85,18 +85,30 @@ function run!(
...
+    max_priority = -Inf
     for (var_id, val) in input.solution
         # Do not consider continuous variables as branching candidates
         getperenkind(master, var_id) == Continuous && continue
-        getbranchingpriority(master, var_id) < input.minimum_priority && continue
+        #getbranchingpriority(master, var_id) < input.minimum_priority && continue
```
Comment (guimarqu) [inline src/Algorithm/branching/varbranching.jl:93]:
```suggestion
```

File: src/Algorithm/branching/varbranching.jl [hunk 11]
```diff
@@ -85,18 +85,30 @@ function run!(
...
         if !isinteger(val, input.int_tol)
-            #description string is just the variable name
-            candidate = VarBranchingCandidate(getname(master, var_id), var_id)
-            local_id += 1
-            push!(groups, BranchingGroup(candidate, local_id, val))
+            brpriority = getbranchingpriority(master, var_id)
+            if (brpriority > max_priority)
```
Comment (guimarqu) [inline src/Algorithm/branching/varbranching.jl:96]:
```suggestion
            if brpriority > max_priority
```

File: src/Algorithm/branching/varbranching.jl [hunk 12]
```diff
@@ -85,18 +85,30 @@ function run!(
...
+                max_priority = brpriority
+                selected_vars = [(var_id, val)]
+            elseif (brpriority == max_priority)
```
Comment (guimarqu) [inline src/Algorithm/branching/varbranching.jl:99]:
```suggestion
            elseif brpriority == max_priority
```

File: src/Algorithm/branching/varbranching.jl [hunk 13]
```diff
@@ -85,18 +85,30 @@ function run!(
...
+        #getbranchingpriority(master, var_id) < input.minimum_priority && continue
         if !isinteger(val, input.int_tol)
-            #description string is just the variable name
-            candidate = VarBranchingCandidate(getname(master, var_id), var_id)
-            local_id += 1
-            push!(groups, BranchingGroup(candidate, local_id, val))
+            brpriority = getbranchingpriority(master, var_id)
+            if (brpriority > max_priority)
+                max_priority = brpriority
+                selected_vars = [(var_id, val)]
```
Comment (guimarqu) [inline src/Algorithm/branching/varbranching.jl:98]: Does it mean that you create a new array each time you find a non-integer variable that has a greater priority than the current greatest priority?

I personally think it's better to loop twice over the solution. A first time to get the greatest branching priority, a second time to fill the `groups` array.

Comment (rrsadykov) [reply src/Algorithm/branching/varbranching.jl]: Ok, I have modified the code in the latest commit as you suggested

**Review by guimarqu — APPROVED** [review]

**Pull request merged**